analysis/syntactic/structinit/structinit.go

```diff
@@ -0,0 +1,552 @@
+		for i := 0; i < st.NumFields(); i++ {
+			field := st.Field(i)
+			if matchStructType(field.Type(), target) { // recursive call
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// findAllocPosition returns the best approximation of instr's position.
+// This is kind of a hack because MakeInterface instructions don't have
+// positions, so this returns the position of the first store instruction that
+// stores the interface value.
+func findAllocPosition(fset *token.FileSet, instr ssa.Instruction) token.Position {
+	if instr.Pos().IsValid() {
+		return fset.Position(instr.Pos())
+	}
+
+	if mk, ok := instr.(*ssa.MakeInterface); ok {
+		for _, ref := range *mk.Referrers() {
+			if s, ok := ref.(*ssa.Store); ok {
+				if s.Val == mk && s.Pos().IsValid() {
+					return fset.Position(s.Pos())
+				}
+			}
+		}
+	}
+
+	panic(fmt.Errorf("invalid instruction type: %T", instr))
+}
+
+type writeToField struct {
+	structTypes structTypes
+	fieldType   *types.Var
+	write       InvalidWrite
+}
+
+func isInvalidWrite(res AnalysisResult, store *ssa.Store, pos token.Position) (writeToField, bool) {
+	field, ok := store.Addr.(*ssa.FieldAddr)
+	if !ok {
+		return writeToField{}, false
+	}
+
+	structTypes, ok := isStoreToStructPtr(field.X)
+	if !ok {
+		return writeToField{}, false
+	}
+
+	infos, ok := res.InitInfos[structTypes.named]
+	if !ok {
+		return writeToField{}, false
+	}
+
+	fieldType := structTypes.strct.Field(field.Field)
+	wantConst, ok := infos.fieldToConst[fieldType]
+	if !ok {
+		// field not in spec
+		return writeToField{}, false
+	}
+
+	gotConst, ok := store.Val.(*ssa.Const)
+	if !ok {
+		return writeToField{}, false
+	}
+
+	if gotConst == nil || gotConst.Value == nil {
+		panic(fmt.Errorf("unexpected nil constant %+v in store instruction to field %v: %v", gotConst, field, store))
+	}
+	// compare the underlying constant values
+	if gotConst.Value == wantConst.Value.Value {
+		return writeToField{}, false
+	}
+
+	return writeToField{
+		structTypes: structTypes,
+		fieldType:   fieldType,
+		write: InvalidWrite{
+			Got:   gotConst.Value,
+			Want:  wantConst.Value.Value,
+			Instr: store,
+			Pos:   pos,
+		},
+	}, true
+}
+
+func isStoreToStructPtr(v ssa.Value) (structTypes, bool) {
+	if ptr, ok := v.Type().(*types.Pointer); ok {
+		if n, ok := ptr.Elem().(*types.Named); ok {
+			if s, ok := n.Underlying().(*types.Struct); ok {
+				return structTypes{strct: s, named: n}, true
+			}
+		}
+	}
+
+	return structTypes{}, false
+}
+
+func findNamedConst(program *ssa.Program, valCi config.CodeIdentifier) (*ssa.NamedConst, bool) {
+	pkgs := program.AllPackages()
+	for _, pkg := range pkgs {
+		for _, mem := range pkg.Members {
+			if c, ok := mem.(*ssa.NamedConst); ok {
+				if valCi.MatchConst(c) && c.Value != nil {
+					return c, true
+				}
+			}
+		}
+	}
+
+	return nil, false
+}
+
+// ReportResults writes res to a string and returns true if the analysis should fail.
+func ReportResults(res AnalysisResult) (string, bool) {
```

Comment:
Will ReportResults have usage outside of logging results? i.e. is not using a logger intentional?

Reply:
I wanted the function to be logger-agnostic since it's called from the CLI but I can change it to use the standard library logger if that would be better

Reply:
That's a good point